src/main/java/org/ice4j/ice/ComponentSocket.java

```diff
@@ -0,0 +1,206 @@
+    private boolean initializedActive = false;
+
+    /**
+     * The {@link Logger} used by {@link MergingDatagramSocket} instances.
+     */
+    private final Logger logger;
+
+    /**
+     * Initializes a new {@link MergingDatagramSocket} instance.
+     * @throws SocketException
+     */
+    ComponentSocket(Component component, Logger levelDelegate)
+        throws SocketException
+    {
+        super(levelDelegate);
+
+        this.component = component;
+        this.logger = new Logger(classLogger, levelDelegate);
+        component.getParentStream().addPairChangeListener(this);
+    }
+
+    /**
+     * {@inheritDoc}
+     * </p>
+     * Verifies that the source of the packet is an authorized remote address.
+     */
+    protected boolean accept(DatagramPacket p)
+    {
+        return authorizedAddresses.contains(p.getSocketAddress());
+    }
+
+    /**
+     * Adds a specific address to the list of authorized remote addresses.
+     * @param address the address to add.
+     */
+    private void addAuthorizedAddress(SocketAddress address)
+    {
+        synchronized (authorizedAddressesSyncRoot)
+        {
+            if (authorizedAddresses.contains(address))
+            {
+                return;
+            }
+
+            logger.info("Adding allowed address: " + address);
+
+            Set<SocketAddress> newSet
+                = Collections.newSetFromMap(
+                        new ConcurrentHashMap<SocketAddress, Boolean>());
+            newSet.addAll(authorizedAddresses);
+            newSet.add(address);
+
+            authorizedAddresses = newSet;
+        }
+    }
+
+    /**
+     * {@inheritDoc}
+     * </p>
+     * Handles property change events coming from ICE pairs.
+     * @param event
+     */
+    @Override
+    public void propertyChange(PropertyChangeEvent event)
+    {
+        String propertyName = event.getPropertyName();
+        if (!(event.getSource() instanceof CandidatePair))
+        {
+            return;
+        }
+
+        CandidatePair pair = (CandidatePair) event.getSource();
+        if (!pair.getParentComponent().equals(component))
+        {
+            // Events are fired by the IceMediaStream, which might have
+            // multiple components. Make sure that we only handle events for
+            // our own component.
+            return;
+        }
+
+        if (IceMediaStream.PROPERTY_PAIR_STATE_CHANGED.equals(propertyName))
+        {
+            CandidatePairState newState
+                = (CandidatePairState) event.getNewValue();
+
+            if (CandidatePairState.SUCCEEDED.equals(newState))
+            {
+                addAuthorizedAddress(
+                        pair.getRemoteCandidate().getTransportAddress());
+            }
+        }
+        else if (IceMediaStream.PROPERTY_PAIR_NOMINATED.equals(propertyName))
+        {
+            if (initializedActive)
+            {
+                return;
+            }
+            initializedActive = true;
```

Comment:
ok, so is it true that only a single thread will go through that code?  otherwise a second thread could enter and make a second call to initializeActive if the first one didn't finish and set the flag yet. this is making me realize that we probably need to either: a) move the if down as well, but then we've potentially allocated some stuff/did some work we're not going to use or b) move the setting of the flag back up.  i just worry with it being up that it's more prone to having an early exit or exception inserted that results in initializeActive never getting called.  maybe that's too paranoid though and it's fine the way it was. :/